spring-integration-mongodb/src/test/java/org/springframework/integration/mongodb/outbound/MongoDbOutboundGatewayTests.java

```diff
@@ -302,6 +305,46 @@ public void testWithCollectionNameExpressionSpecified() throws Exception {
 		assertEquals("anotherCollection", collectionNameExpression.getValue());
 	}
 
+	@Test
+	@MongoDbAvailable
+	public void testWithCollectionCallbackCount() throws Exception {
+		Message<String> message = MessageBuilder.withPayload("").build();
+		MongoDbOutboundGateway gateway = createGateway();
+		gateway.setEntityClass(Person.class);
+		gateway.setExpectSingleResult(true);
+		gateway.setCollectionNameExpression(new LiteralExpression("data"));
+
+		gateway.setCollectionCallback(MongoCollection::count);
+		gateway.afterPropertiesSet();
+
+		long result = (long) gateway.handleRequestMessage(message);
+
+		assertEquals(4, result);
+	}
+
+	@Test
+	@MongoDbAvailable
+	public void testWithCollectionCallbackFindOne() throws Exception {
+		Message<String> message = MessageBuilder.withPayload("").build();
+		MongoDbOutboundGateway gateway = createGateway();
+		gateway.setEntityClass(Person.class);
+		gateway.setExpectSingleResult(true);
```

Comment:
There must be some not on this setter that the option makes sense only if `query` is specified.
OTOH it would be better to also add assert to the `MongoDbOutboundGateway . doInit()` that `collectionCallback` and `query` are mutually exclusive.